### PR TITLE
API SecureManager::SendMessage, use SecureSessoinHandle instead NodeId

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
Currently `SecureSessionMgrBase::SendMessage` takes a `PeerNodeId` as key to find proper `PeerConnectionState`, but is way may causing unexpected behavior. For example.
 * If there are multiple connection to the same node, there is no way to specify which connection to use,
 * When switching a connection (SPAKE -> SIGMA, or SIGMA -> SPAKE), the application can not aware the change, and mistakenly send messages to wrong channel.
 * The ACL permission information is associated with a certification, which is used to initialized the connection, the connection derives access permission from it. So we should use connection as key to verify permission. Using node id is insecure and may cause potential security flaws.

 #### Summary of Changes
1. `SecureSessionMgrBase::SendMessage` take a `Transport::PeerConnectionState *` instead of `PeerNodeId`
2. Enable `SecureSessionMgrDelegate::OnNewConnection` so that upper components can receive a `PeerConnectionState` object
3. Add `SecureSessionMgrDelegate::OnConnectionExpired` callback, so upper components can acknowledge that the connection has been released.

### CODE REVIEW

**Comment by kghost** [comment]
Before I resolve all CI failures, I want to collect opinions first.

**Comment by andy31415** [comment]
I have conmcerns about the approach: I agree with the intent that there may be more communication paths than 1 to a node, which we plan to call 'exchanges'. The PeerConnectionState however was meant to encapsulate things like encryption keys and counters, which applications should not see.

I would rather imagine we would like something like a 'connectionID' exposed that could be something like a pair<nodeid, exchangeid>. This sounds like a minimal information required rather than an entire large structure pointer.

**Comment by andy31415** [comment]
> I have conmcerns about the approach: I agree with the intent that there may be more communication paths than 1 to a node, which we plan to call 'exchanges'. The PeerConnectionState however was meant to encapsulate things like encryption keys and counters, which applications should not see.
> 
> I would rather imagine we would like something like a 'connectionID' exposed that could be something like a pair<nodeid, exchangeid>. This sounds like a minimal information required rather than an entire large structure pointer.

Marked as 'changes requested' since I believe the PeerConnectionState to be too large of an object to pass around and ideally API boundaries should be with smaller pieces of data.

**Comment by pan-apple** [comment]
I would also like to understand the scenario where the user of the secure session manager is aware of SPAKE based connection vs Sigma based connection. Currently, SPAKE based secure session is only used by Rendezvous state machine. It's not exposed to the controller or the ZCL layer.

The multiple peer connections between two nodes are supported, but mainly for

1. Unicast/Multicast keys
2. Key rolling (where the old key/session is being torn down and new session is being established)

The first could be disambiguated by adding a multicast API to the secure session manager.
The second should be self contained inside the secure session layer.

**Comment by kghost** [comment]
@andy31415 
> I would rather imagine we would like something like a 'connectionID' exposed that could be something like a pair<nodeid, exchangeid>. This sounds like a minimal information required rather than an entire large structure pointer.

> Marked as 'changes requested' since I believe the PeerConnectionState to be too large of an object to pass around and ideally API boundaries should be with smaller pieces of data.


I agree with this part, an `ExchangeContext` could be enough for application, node id can be retrieved via the context. But here I'm proposing the internal relation ship between `ExchangeContext` and `SecureSession`. The `Transport::PeerConnectionState *` is only exposed to the `ExchangeManager`, not application.

@pan-apple 
> I would also like to understand the scenario where the user of the secure session manager is aware of SPAKE based connection vs Sigma based connection. Currently, SPAKE based secure session is only used by Rendezvous state machine. It's not exposed to the controller or the ZCL layer.

Problem 1 and 2 may not be a real problem, but 3 is. Lets focus on the changes.,

How to implement `ExchangeContext::GetPeerCertification` ? The ACL information is stored inside the certification, and applications need it to grant access permission. Using only node id to look up peer state may not be a good idea, it is kind of tricky and unstable.

**Comment by pan-apple** [comment]
> @andy31415
> 
> > I would rather imagine we would like something like a 'connectionID' exposed that could be something like a pair<nodeid, exchangeid>. This sounds like a minimal information required rather than an entire large structure pointer.
> 
> > Marked as 'changes requested' since I believe the PeerConnectionState to be too large of an object to pass around and ideally API boundaries should be with smaller pieces of data.
> 
> I agree with this part, an `ExchangeContext` could be enough for application, node id can be retrieved via the context. But here I'm proposing the internal relation ship between `ExchangeContext` and `SecureSession`. The `Transport::PeerConnectionState *` is only exposed to the `ExchangeManager`, not application.
> 
> @pan-apple
> 
> > I would also like to understand the scenario where the user of the secure session manager is aware of SPAKE based connection vs Sigma based connection. Currently, SPAKE based secure session is only used by Rendezvous state machine. It's not exposed to the controller or the ZCL layer.
> 
> Problem 1 and 2 may not be a real problem, but 3 is. Lets focus on the changes.,
> 
> How to implement `ExchangeContext::GetPeerCertification` ? The ACL information is stored inside the certification, and applications need it to grant access permission. Using only node id to look up peer state may not be a good idea, it is kind of tricky and unstable.

What value (besides peer node ID) would be needed to look up the peer certificates? Is it peer's encryption key ID.

**Comment by rwalker-apple** [comment]
@pan-apple, still looking at this?

**Comment by kghost** [comment]
I'm looking for suggestions to move forward. Both `Tuple<NodeId, KeyId>` and direct pointer can work. If we can decide to move this direction, we can compare them deeply.

**Comment by pan-apple** [comment]
> I'm looking for suggestions to move forward. Both `Tuple<NodeId, KeyId>` and direct pointer can work. If we can decide to move this direction, we can compare them deeply.

I don't have a strong opinion on either approach. Exposing the `PeerConnectionState` seems a bit excessive. How about we start with `Tuple<NodeId, KeyId>` and see if any other information is needed and add in future?

**Comment by kghost** [comment]
I temporary use `std::tuple` to illustrate.

The `tuple` way has several pitfalls:
  1. it is 3 times larger than the pointer, which will increase memory footprint and hurt performance.
  2. `SecureSessoinMgr` is needed to convert the `Tuple<NodeId, KeyId>` to `PeerConnectionState` object, applications above `ExchangeManage` may have difficulty to access `SecureSessoinMgr`. Without access to `SecureSessoinMgr` applications may not be able to get information in  `PeerConnectionState` like `PeerAddress`, `PeerCertification`, and so on.
  3. it is possible that the id number wrap around and causing a conflict with a dangling tuple whose associated key has be released long ago. A malicious attacker can set up an attack which eat up KeyId space, and construct a specific KeyId to attack an flawed application which forget to release the tuple.

Using the `PeerConnectionState` pointer introduces the problem that its lifecycle must be carefully managed, If the object has been released without notify who uses and stores the pointer, can create a dangling pointer. It can be overcome by keeping it inside `ExchangeContext` and preventing applications from accessing the pointer. We will handle the problem inside `ExchangeManager`, and expose `ExchangeContext` interface to applications. `ExchangeContext::GetPeerState` will return a reference instead of pointer, then the application will not be able to create a dangling pointer. `ExchangeContext` will be always available to applications.

**Comment by rwalker-apple** [comment]
> I temporary use `std::tuple` to illustrate.
> 
> The `tuple` way has several pitfalls:
> 
> 1. it is 3 times larger than the pointer, which will increase memory footprint and hurt performance.
> 2. `SecureSessoinMgr` is needed to convert the `Tuple<NodeId, KeyId>` to `PeerConnectionState` object, applications above `ExchangeManage` may have difficulty to access `SecureSessoinMgr`. Without access to `SecureSessoinMgr` applications may not be able to get information in  `PeerConnectionState` like `PeerAddress`, `PeerCertification`, and so on.
> 3. it is possible that the id number wrap around and causing a conflict with a dangling tuple whose associated key has be released long ago. A malicious attacker can set up an attack which eat up KeyId space, and construct a specific conflict KeyId to attack an flawed application which forget to release the tuple.
> 
> Using the `PeerConnectionState` pointer introduces the problem that its lifecycle must be carefully managed, If the object has been released without notify who uses and stores the pointer, can create a dangling pointer. It can be overcome by keeping it inside `ExchangeContext` and preventing applications from accessing the pointer. We will handle the problem inside `ExchangeManager`, and expose `ExchangeContext` interface to applications. `ExchangeContext::GetPeerState` will return a reference instead of pointer, then the application will not be able to create a dangling pointer. `ExchangeContext` will be always available to applications.

@pan-apple ?

**Comment by kghost** [comment]
@pan-apple Any progress ?

**Comment by kghost** [comment]
@rwalker-apple  @pan-apple Would you please check this PR, using only `PeerNodeId` to send messages is not right. It become harder and harder to change the API when more feature which is depending on this API are developed. We should fix the API ASAP.

File: src/controller/CHIPDevice.cpp [hunk 1]
```diff
@@ -69,7 +69,7 @@ CHIP_ERROR Device::SendMessage(System::PacketBuffer * buffer)
         resend = buffer;
     }
 
-    err    = mSessionManager->SendMessage(mDeviceId, buffer);
+    err    = mSessionManager->SendMessage(mSecureSession, buffer);
```
Comment (pan-apple) [inline src/controller/CHIPDevice.cpp:72]: @kghost , how does the value of `mSecureSession` get populated when the device is already paired, and it is `deserialized` from the persistent storage?

File: src/controller/CHIPDeviceController.cpp [hunk 2]
```diff
@@ -309,24 +309,57 @@ CHIP_ERROR DeviceController::ServiceEventSignal()
...
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint16_t index = 0;
+
+    VerifyOrExit(mState == State::Initialized, err = CHIP_ERROR_INCORRECT_STATE);
+
+    index = FindDeviceIndex(mgr->GetPeerConnectionState(session)->GetPeerNodeId());
+    VerifyOrExit(index < kNumMaxActiveDevices, err = CHIP_ERROR_INVALID_DEVICE_DESCRIPTOR);
+
+    mActiveDevices[index].OnNewConnection(session, mgr);
```
Comment (pan-apple) [inline src/controller/CHIPDeviceController.cpp:322]: This makes the device object use the secure session corresponding to the keyID that was derived during pairing (which is based on SPAKE2P handshake). How will it switch to using the key derived by SIGMA handshake (which is more secure, and should typically be used for all communication once pairing process is complete).

File: src/controller/CHIPDevice.cpp (hunk 1 above)
Comment (kghost) [reply src/controller/CHIPDevice.cpp:72]: I'm not sure `SecureSessionHandle` should be able to be serialized to persistent storage. There will be several ways to obtain a handle.

Messaging layer:
1. Server: From `SecureSessoinDelegate::OnNewConnection`
2. Client: Not sure, while looking the newest code, `SecurePairingSession::DeriveSecureSession` takes my attention

Application interface (Messaging layer API):
1. Server: Register unsolicited message handler, then from `ExchangeDelegate::OnMessageReceived` of that handler
2. Client: Using `Channel` API, https://github.com/project-chip/connectedhomeip/pull/4019

File: src/controller/CHIPDeviceController.cpp (hunk 2 above)
Comment (kghost) [reply src/controller/CHIPDeviceController.cpp:322]: Maybe I'm not quite understand your question, but IMHO, `SecureSessionHandle` can represent both `SPAKE2` and `SIGMA` session. And there can be multiple connections to a device, `OnNewConnection` is triggered for every connection.

If the `Device` class wants to "persistent" a connection, it can choose one connection it wants.

File: src/controller/CHIPDevice.cpp (hunk 1 above)
Comment (pan-apple) [reply src/controller/CHIPDevice.cpp]: I suspect, the current PR will break the communication between the controller and the device. After pairing is complete, the controller saves the device information in the persistent storage. When a message is sent to the device, the controller creates the `CHIPDevice` object by reading it from the storage. With the current PR, the created (deserialized) device object will not have a valid `mSecureSession` object.

Will you be making more changes in the current PR to fix this issue?

File: src/controller/CHIPDeviceController.cpp (hunk 2 above)
Comment (pan-apple) [reply src/controller/CHIPDeviceController.cpp:347]: Sorry, let me rephrase the question. It's two parts.

1. What's the use case where an application decides to use a `PASE` based connection (established by SPAKE2P), after a `CASE` based connection (established via SIGMA) is already established?
2. How would the application know which KeyID corresponds to which connection. (it's ok if this logic is added in a follow on PR).

File: src/controller/CHIPDevice.cpp (hunk 1 above)
Comment (kghost) [reply src/controller/CHIPDevice.cpp]: There will be a Channel API to create the `mSecureSession` object. But Channel implementation depends on this PR. So to break to loop, we can temporary create an `mSecureSession` from `PeerNodeId`.

File: examples/common/chip-app-server/Server.cpp [hunk 3]
```diff
@@ -61,23 +61,13 @@ class ServerCallback : public SecureSessionMgrDelegate
...
         VerifyOrExit(!buffer.IsNull(), ChipLogProgress(AppServer, "Received data but couldn't process it..."));
         VerifyOrExit(header.GetSourceNodeId().HasValue(), ChipLogProgress(AppServer, "Unknown source for received message"));
-
-        VerifyOrExit(state->GetPeerNodeId() != kUndefinedNodeId, ChipLogProgress(AppServer, "Unknown source for received message"));
-
-        state->GetPeerAddress().ToString(src_addr, sizeof(src_addr));
-
-        ChipLogProgress(AppServer, "Packet received from %s: %zu bytes", src_addr, static_cast<size_t>(data_len));
-
-        HandleDataModelMessage(header, std::move(buffer), mgr);
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/Server.cpp:80]: Maybe switch this code to reporting the node id instead of the source address (which it can get from the `SecureSessionHandle` , presumably, at least for a CASE session)?  Either way, keeping the "we got N bytes logging" would be useful.

File: src/controller/CHIPDevice.cpp [hunk 4]
```diff
@@ -175,8 +175,19 @@ CHIP_ERROR Device::Deserialize(const SerializedDevice & input)
     return error;
 }
 
+void Device::OnNewConnection(SecureSessionHandle session, SecureSessionMgr * mgr)
+{
+    mState         = ConnectionState::SecureConnected;
+    mSecureSession = session;
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:181]: Do we expect `SecureSessionHandle` to always be copyable?  Or are we likely to switch to move semantics on it at some point?  Put another way, should there be an `std::move()` here?

File: src/controller/CHIPDevice.cpp [hunk 5]
```diff
@@ -175,8 +175,19 @@ CHIP_ERROR Device::Deserialize(const SerializedDevice & input)
...
+}
+
+void Device::OnConnectionExpired(SecureSessionHandle session, SecureSessionMgr * mgr)
+{
+    mState = ConnectionState::NotConnected;
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:186]: Should `mSecureSession` be reset in some way?

File: src/controller/CHIPDevice.h [hunk 6]
```diff
@@ -156,6 +156,26 @@ class DLL_EXPORT Device
      **/
     CHIP_ERROR Deserialize(const SerializedDevice & input);
 
+    /**
+     * @brief
+     *   Called when a new pairing is being established
+     *
+     * @param state   connection state
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:163]: There is no such param

File: src/controller/CHIPDevice.h [hunk 7]
```diff
@@ -156,6 +156,26 @@ class DLL_EXPORT Device
...
+     */
+    void OnNewConnection(SecureSessionHandle session, SecureSessionMgr * mgr);
+
+    /**
+     * @brief
+     *   Called when a connection is closing.
+     *
+     *   The receiver should release all resources associated with the connection.
+     *
+     * @param state   connection state
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:174]: No such param...

File: src/controller/CHIPDevice.h [hunk 8]
```diff
@@ -191,6 +213,8 @@ class DLL_EXPORT Device
 
     NodeId GetDeviceId() const { return mDeviceId; }
 
+    SecureSessionHandle GetSecureSession() const { return mSecureSession; }
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:216]: What will this return if `OnNewConnection` has not happened yet?  Will it even be initialized memory?

File: src/controller/CHIPDeviceController.cpp [hunk 9]
```diff
@@ -395,6 +429,21 @@ void DeviceController::ReleaseAllDevices()
     }
 }
 
+uint16_t DeviceController::FindDeviceIndex(SecureSessionHandle session)
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:432]: I know this is what the code used to look like with node id, but is there a reason this is not `FindDevice` returning `Device *`, with `nullptr` returned if not found?  Then we could use a ranged for loop instead of the indexing we are doing now, and callers wouldn't need to index back into the array; they would just use the pointer they got, if not null.

File: examples/common/chip-app-server/Server.cpp (hunk 3 above)
Comment (kghost) [reply examples/common/chip-app-server/Server.cpp:80]: The member of the `SecureSessionHandle` is not decided yet. I don't want to expose the internal detail of `SecureSessionHandle` class.

I'm still considering what do we expose from session interface.

```c++
auto session = mSecureSessionManager->GetSession(handle /* an secure session handle */);
bool isPaseSession = session->IsPaseSession(); // or session->IsCaseSession() for case check.
auto nodeid = session->GetPeerNodeId();
auto localKey = session->GetLocalKeyId();
auto peerKey = session->GetPeerKeyId();
```

File: src/controller/CHIPDevice.cpp (hunk 4 above)
Comment (kghost) [reply src/controller/CHIPDevice.cpp:181]: Yes, I'll expect that the handle is POD, and always be copyable. In the future, only messaging layer will use the handle, the handle class won't be exposed to applications or controller API.

File: src/controller/CHIPDevice.h (hunk 8 above)
Comment (kghost) [reply src/controller/CHIPDevice.h:216]: I have removed `GetSecureSession`.

File: src/controller/CHIPDeviceController.cpp (hunk 9 above)
Comment (kghost) [reply src/controller/CHIPDeviceController.cpp:432]: While audit the usage of `FindDeviceIndex`, I don't think there is any reason preventing it from returning a `Device *`.

We can refactor it in another PR. but here I want to keep it consistent with `uint16_t DeviceController::FindDeviceIndex(NodeId id)`.

File: src/controller/CHIPDevice.h (hunk 7 above)
Comment (kghost) [reply src/controller/CHIPDevice.h:174]: done

File: src/controller/CHIPDevice.h (hunk 6 above)
Comment (kghost) [reply src/controller/CHIPDevice.h:163]: done

File: src/controller/CHIPDevice.cpp (hunk 5 above)
Comment (kghost) [reply src/controller/CHIPDevice.cpp:186]: done

File: examples/common/chip-app-server/Server.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply examples/common/chip-app-server/Server.cpp:80]: That's fair, but we really will need a way to get the peer node ID for CASE sessions.  And we really should keep the length logging here.

File: src/controller/CHIPDeviceController.cpp (hunk 9 above)
Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:432]: Yep, followup is fine.  Just file an issue, please.

File: src/controller/CHIPDevice.h [hunk 10]
```diff
@@ -191,6 +213,8 @@ class DLL_EXPORT Device
 
     NodeId GetDeviceId() const { return mDeviceId; }
 
+    bool MatchSession(SecureSessionHandle session) const { return mSecureSession == session; }
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:216]:
```suggestion
    bool MatchesSession(SecureSessionHandle session) const { return mSecureSession == session; }
```

File: src/transport/SecureSessionMgr.h [hunk 11]
```diff
@@ -44,6 +44,23 @@ namespace chip {
 
 class SecureSessionMgr;
 
+
+class SecureSessionHandle
+{
+public:
+    SecureSessionHandle() : mPeerNodeId(kAnyNodeId), mPeerKeyId(0) {}
+    SecureSessionHandle(NodeId peerNodeId, uint16_t peerKeyId) : mPeerNodeId(peerNodeId), mPeerKeyId(peerKeyId) {}
```
Comment (bzbarsky-apple) [inline src/transport/SecureSessionMgr.h:52]: Should this constructor be protected, so only `SecureSessionMgr` can use it?

Comment (kghost) [reply src/transport/SecureSessionMgr.h:52]: In the future, it is. But for now, InteractionModel need to create it using this one.

File: examples/common/chip-app-server/Server.cpp (hunk 3 above)
Comment (kghost) [reply examples/common/chip-app-server/Server.cpp:80]: Good, I'll make it so.

File: src/controller/CHIPDeviceController.cpp (hunk 2 above)
Comment (kghost) [reply src/controller/CHIPDeviceController.cpp:347]: I have created another PR to address the problem (https://github.com/project-chip/connectedhomeip/pull/4171), Let's continue the discussion on that PR.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by kghost** [comment]
@pan-apple  @andy31415 this one is ready for merge, please do a final check.

In the future, I may need to change the member if `SecureSesoinHandle` class, since the internal of this class is never exposed, it should be a very easy task.

**Review by rwalker-apple — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by kghost** [comment]
This one is ready for merge

**Comment by mspang** [comment]
> This one is ready for merge

@kghost conflict

**Comment by rwalker-apple** [comment]
@jelderton @hawk248 @saurabhst @BroderickCarlin ?

**Review by saurabhst — APPROVED** [review]

**Comment by yufengwangca** [comment]
This change seems break the end-to-end messaging testing which is temporarily disabled in CI.

With this change, the app does not specify the target NodeID, but relay on the following callback to update the SecureSessionHandle.

    void OnNewConnection(chip::SecureSessionHandle session, chip::SecureSessionMgr * mgr) override { 
        mSecureSession = session;
    }

We can only specify one delegate of SecureSessionMgr at a time, but there are multiple connections managed by the SecureSessionMgr, how we make sure the SecureSessionMgr can pass the generated SecureSessionHandle to the corresponding target of its connection?

For example, there are multiple objects are set as the delegate of  SecureSessionMgr,  including DeviceController, ExchangeManager and application itself. The later SetDelegate() will overwrite the previous one.  

Echo Application:
    gSessionManager.SetDelegate(&gTestSecureSessionMgrDelegate);   // Set the echo app as the delegate of SecureSessionMgr.
    err = gExchangeManager.Init(&gSessionManager);  // Set the ExchangeManager as the delegate of SecureSessionMgr.

The echo application won't received the generated SecureSessionHandle  and it will sendMessage to garbage target.

**Comment by yufengwangca** [comment]
Register https://github.com/project-chip/connectedhomeip/issues/4451 to track this issue.
